library/spdm_common_lib/libspdm_com_crypto_service.c

```diff
@@ -517,54 +581,72 @@ uint8_t libspdm_get_cert_slot_count(libspdm_context_t *spdm_context)
 /**
  * This function verifies the digest.
  *
- * @param  spdm_context                  A pointer to the SPDM context.
- * @param  digest                       The digest data buffer.
- * @param  digest_count                   size of the digest data buffer.
+ * @param  spdm_context                A pointer to the SPDM context.
+ * @param  digest                      The digest data buffer.
+ * @param  slot_mask                   the mask indicates slot id.
  *
  * @retval true  digest verification pass.
  * @retval false digest verification fail.
  **/
 bool libspdm_verify_peer_digests(libspdm_context_t *spdm_context, const void *digest,
-                                 size_t digest_count)
+                                 uint8_t slot_mask)
 {
     size_t hash_size;
     const uint8_t *hash_buffer;
+    uint8_t slot_id;
+#if LIBSPDM_RECORD_TRANSCRIPT_DATA_SUPPORT
     uint8_t cert_chain_buffer_hash[LIBSPDM_MAX_HASH_SIZE];
     const uint8_t *cert_chain_buffer;
     size_t cert_chain_buffer_size;
-    size_t index;
     bool result;
+#endif
 
-    cert_chain_buffer = spdm_context->local_context.peer_cert_chain_provision;
-    cert_chain_buffer_size = spdm_context->local_context.peer_cert_chain_provision_size;
-    if ((cert_chain_buffer != NULL) && (cert_chain_buffer_size != 0)) {
-        hash_size = libspdm_get_hash_size(spdm_context->connection_info.algorithm.base_hash_algo);
-        hash_buffer = digest;
+    hash_size = libspdm_get_hash_size(spdm_context->connection_info.algorithm.base_hash_algo);
+    hash_buffer = digest;
 
-        result = libspdm_hash_all(
-            spdm_context->connection_info.algorithm.base_hash_algo,
-            cert_chain_buffer, cert_chain_buffer_size,
-            cert_chain_buffer_hash);
-        if (!result) {
-            LIBSPDM_DEBUG((LIBSPDM_DEBUG_INFO,
-                           "!!! verify_peer_digests - FAIL (hash calculation) !!!\n"));
-            return false;
-        }
+    for (slot_id = 0; slot_id < SPDM_MAX_SLOT_COUNT; slot_id++) {
+        if (slot_mask & (1 << slot_id)) {
+#if LIBSPDM_RECORD_TRANSCRIPT_DATA_SUPPORT
+            if (spdm_context->connection_info.peer_used_cert_chain[slot_id].buffer_size != 0) {
```

Comment:
Originally, this API is to check the runtime data v.s. local_context.

But, the local_context.peer_cert_chain_provision is moved. The check is change to runtime data v.s. connection_info.

I dont believe that is the expectation.

I recommend we remove libspdm_verify_peer_digests() API at all, because there is nothing to verify.

Reply:
In my opinion the libspdm_verify_peer_digests() was supposed to be checking runtime retrieved digests data.
Refer to SPDM specification v1.2 `Figure 8 — Responder authentication: Example certificate retrieval flow`, 
> 2. Compare digests in DIGESTS response
message to cached digests. _**Continue if
no match is found**_.

The word **_Continue_** here in my understanding is to perform `GET_CERTIFICATE` afterwards.
Thus as I see the use cases of libspdm_verify_peer_digests() are still there:
1. First time `GET_DIGESTS`. The libspdm_verify_peer_digests() should return FALSE (since there is nothing in local context to verify with the retrieved digests), indicate people to perform `GET_CERTIFICATE` (Continue if
no match is found), and cache digest of certificate chain per slot in local context connection_info by the `GET_CERTIFICATE`.
2. `GET_DIGESTS` after `LIBSPDM_CONNECTION_STATE_AUTHENTICATED`. The libspdm_verify_peer_digests() should return actual check result (runtime retrieved digests data v.s. cached digest per slot in local context connection_info), to tell if the peer identity is changed (after a `SET_CERTIFICATE`) or tampered, and a `GET_CERTIFICATE` afterwards may be needed base upon the check result.

Currently `GET_CERTIFICATE` is called only if `GET_DIGESTS` is succeeded so I keep the libspdm_verify_peer_digests() to return TRUE at first time it is called (nothing is cached in local context connection_info). It can be filed to another issue and be resolved later.
https://github.com/DMTF/spdm-emu/blob/9202e51b57abba044387746a926832df8174114d/spdm_emu/spdm_requester_emu/spdm_requester_authentication.c#L42-L54

In summary, I think libspdm_verify_peer_digests() is needed.

Reply:
> First time GET_DIGESTS. The libspdm_verify_peer_digests() should return FALSE

This is not the purpose, when we designed libspdm_verify_peer_digests().
The expectation of verify_xxx to check if the data from the peer meets the basic requirement before return the consumer.
As such, we only verify if it matches the provisioned data. If no data is provisioned, this API should just return TRUE, because there is nothing to verify.

> GET_DIGESTS after LIBSPDM_CONNECTION_STATE_AUTHENTICATED. The libspdm_verify_peer_digests() should return actual check result

This is also not the purpose of verify_xxx API. Thinking of runtime update use case, the change of digest is a valid condition. That helps the consumer to know that the cached digest is no longer valid. The requester should send get_certificate again to get the latest result. 



To me, what you describe above is something like is_cached_data_valid() for runtime data, instead of verify_xxx() according to the provisioned data.
My suggestion is: Don't mix them in raw public key provision patch. Just remove libspdm_verify_peer_digests() in this patch. If you really want to add is_cached_data_valid(), let's use another issue and PR.

Reply:
This explanation helps a lot, thank you.
Now I fully understand the design of `libspdm_verify_peer_digests()` is only to verify if the retrieved digest matches provisioned data before return it to consumer.
I will remove `libspdm_verify_peer_digests()` by next patch set as you suggested since it is no longer needed.

Also, there's not much need to do such `is_cached_data_valid()` thing since it would be consumer's call to send get_certificate (or not) afterwards.